zhmcclient/_os_console.py

```diff
@@ -0,0 +1,378 @@
+
+        Raises:
+
+          zhmcclient.Error: Issue when creating the WebSocket on the HMC.
+          OSConsoleConnectedError: Already connected to the OS console.
+          OSConsoleWebSocketError: Issue when connecting via the WebSocket.
+        """
+        if self.ws is not None:
+            cpc = self.partition.manager.parent
+            raise OSConsoleConnectedError(
+                "Already connected to the OS console for partition "
+                f"{cpc.name}.{self.partition.name}")
+
+        self.ws_uri = self.partition.create_os_websocket()
+        # may raise zhmcclient.Error
+
+        session = self.partition.manager.session
+        self.ws_full_uri = \
+            f"wss://{session.actual_host}:{DEFAULT_HMC_PORT}{self.ws_uri}"
+
+        sslopt = {}
+        if isinstance(session.verify_cert, str):
+            sslopt["cert_reqs"] = ssl.CERT_REQUIRED
+            sslopt["ca_cert_path"] = session.verify_cert
+        elif session.verify_cert is True:
+            sslopt["cert_reqs"] = ssl.CERT_REQUIRED
+            sslopt["ca_cert_path"] = certifi.where()
+        else:
+            sslopt["cert_reqs"] = ssl.CERT_NONE
+        self.ws = websocket.WebSocket(sslopt=sslopt)
+
+        try:
+            self.ws.connect(self.ws_full_uri, timeout=self.ws_timeout)
+        except (websocket.WebSocketException, OSError) as exc:
+            raise OSConsoleWebSocketError(str(exc))
+
+        self.previous_text = None
+
+    def disconnect(self):
+        """
+        Disconnect from the OS console.
+
+        This object must currently be connected to the OS console.
+
+        Raises:
+
+          OSConsoleNotConnectedError: Not connected to the OS console.
+          OSConsoleWebSocketError: Issue when closing the WebSocket.
+        """
+        if self.ws is None:
+            raise OSConsoleNotConnectedError(
+                "Not connected to the OS console")
+
+        try:
+            self.ws.close()
+        except (websocket.WebSocketException, OSError) as exc:
+            raise OSConsoleWebSocketError(str(exc))
+
+        self.ws = None
+        self.ws_uri = None
+        self.ws_full_uri = None
+        self.previous_text = None
+
+    def recv_all(self):
+        """
+        Receive all text data from the OS console until no more data is
+        available within the WebSocket timeout, and return that.
+
+        This object must currently be connected to the OS console.
+
+        In the string that is returned, any sequence of one or more CR (U+000D)
+        characters followed by NL (U+000A) has been translated to NL (i.e. the
+        CR characters are removed).
+
+        Returns:
+
+          str: Received text data, as a Unicode string.
+
+        Raises:
+
+          OSConsoleNotConnectedError: Not connected to the OS console.
+          OSConsoleWebSocketError: Issue when reading data from the WebSocket.
+        """
+        if self.ws is None:
```

Comment:
code duplication: check + raise could go into a dedicated helper method "checkConnected()"

Reply:
DONE
Added `_check_connected()`, and for symmetry also `_check_not_connected()`.